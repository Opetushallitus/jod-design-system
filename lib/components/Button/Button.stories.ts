import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { Button } from './Button';

const meta = {
  title: 'Primitives/Button',
  component: Button,
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Base: Story = {
  args: {
    label: 'Base Button',
    variant: 'base',
    onClick: fn(),
  },
};

export const Primary: Story = {
  args: {
    label: 'Primary Button',
    variant: 'primary',
    onClick: fn(),
  },
};

export const Outlined: Story = {
  args: {
    label: 'Outlined Button',
    variant: 'outlined',
    onClick: fn(),
  },
};

export const Text: Story = {
  args: {
    label: 'Text Button',
    variant: 'text',
    onClick: fn(),
  },
};
