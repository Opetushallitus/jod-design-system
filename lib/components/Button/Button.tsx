import { cx } from '../../cva';

export interface ButtonProps {
  /** Form ID to associate the button with */
  form?: string;
  /** Text shown on the button */
  label: string;
  /** Callback fired on tap/click of the button */
  onClick?: () => void;
  /** Size of the button */
  size?: 'sm' | 'md' | 'lg';
  /** Button variant */
  variant?: 'gray' | 'white' | 'gray-delete' | 'white-delete';
  /** Button disabled for any actions */
  disabled?: boolean;
  /** Icon shown on the button */
  icon?: React.ReactNode;
  /** Icon side */
  iconSide?: 'left' | 'right';
}

const getSizeClassName = (size: ButtonProps['size']) => {
  return cx({
    'ds-text-button-sm ds-px-6 ds-rounded-[30px]': size === 'sm',
    'ds-text-button-md ds-px-6 ds-rounded-[30px]': size === 'md',
    'ds-text-button-lg ds-px-[42px] ds-rounded-[50px]': size === 'lg',
  });
};

const getIconClassName = (size: ButtonProps['size'], leftIcon: boolean, rightIcon: boolean) => {
  return cx({
    'ds-pl-4': (size === 'sm' || size === 'md') && leftIcon,
    'ds-pl-[21px]': size === 'lg' && leftIcon,
    'ds-pr-4': (size === 'sm' || size === 'md') && rightIcon,
    'ds-pr-[21px]': size === 'lg' && rightIcon,
  });
};

const getVariantClassName = (variant: ButtonProps['variant'], disabled: ButtonProps['disabled']) => {
  return cx({
    'ds-text-black ds-bg-bg-gray': variant === 'gray',
    'ds-text-black ds-bg-white': variant === 'white',
    'ds-text-alert ds-bg-bg-gray hover:ds-text-alert active:ds-text-white focus-visible:ds-text-alert':
      variant === 'gray-delete',
    'ds-text-alert ds-bg-white hover:ds-text-alert active:ds-text-white focus-visible:ds-text-alert':
      variant === 'white-delete',
    'active:ds-bg-alert focus-visible:ds-outline-alert': variant?.includes('-delete') && !disabled,
  });
};

const getDisabledClassName = (disabled: ButtonProps['disabled']) => {
  return disabled === false
    ? 'hover:ds-text-accent focus-visible:ds-text-accent active:ds-bg-accent focus-visible:ds-outline-accent focus-visible:ds-outline focus-visible:ds-outline-[3px] focus-visible:ds-outline-offset-[1.5px] active:ds-text-white active:ds-outline-0'
    : 'disabled:ds-text-inactive-gray disabled:ds-cursor-not-allowed';
};

const getButtonClassName = ({
  size,
  variant,
  leftIcon,
  rightIcon,
  disabled,
}: Partial<ButtonProps> & { leftIcon: boolean; rightIcon: boolean }) => {
  return `ds-flex ds-items-center ds-gap-4 ds-select-none ds-group
    ${getSizeClassName(size)}
    ${getIconClassName(size, leftIcon, rightIcon)}
    ${getVariantClassName(variant, disabled)}
    ${getDisabledClassName(disabled)}
  `
    .replace(/\s+/g, ' ')
    .trim();
};

/** Button component for user actions. */
export const Button = ({
  form,
  label,
  onClick,
  size = 'md',
  variant = 'gray',
  disabled = false,
  icon,
  iconSide,
}: ButtonProps) => {
  const leftIcon = icon !== undefined && iconSide === 'left';
  const rightIcon = icon !== undefined && iconSide === 'right';
  const onlyIcon = icon && !leftIcon && !rightIcon;
  const buttonClassName = getButtonClassName({ size, variant, leftIcon, rightIcon, disabled });

  const spanClassName = cx({
    'group-hover:ds-underline group-active:ds-no-underline group-focus-visible:ds-no-underline': !disabled,
    'ds-py-3': size === 'sm',
    'ds-py-[10px]': size === 'md',
    'ds-py-[20px]': size === 'lg',
  });

  return (
    <button
      form={form}
      disabled={disabled}
      type={onClick ? 'button' : 'submit'}
      onClick={onClick}
      className={buttonClassName}
    >
      {leftIcon && <>{icon}</>}
      {onlyIcon ? <>{icon}</> : <span className={spanClassName}>{label}</span>}
      {rightIcon && <>{icon}</>}
    </button>
  );
};
