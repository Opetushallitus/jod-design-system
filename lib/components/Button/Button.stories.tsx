import type { StoryObj } from '@storybook/react-vite';
import { MdArrowBack, MdArrowForward, MdPerson } from 'react-icons/md';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';

import { Button } from './Button';

const meta = {
  title: 'Buttons/Button',
  component: Button,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

const backgrounds = {
  default: 'jod-bg-white',
  values: [{ name: 'jod-bg-white', value: '#fff' }],
};

export const Gray: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5869&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a gray button component for triggering an action.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Muokkaa',
    onClick: fn(),
    variant: 'gray',
  },
};

export const White: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5829&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a white button component for triggering an action.',
      },
    },
  },
  args: {
    label: 'Muokkaa',
    onClick: fn(),
    variant: 'white',
  },
};

export const GrayDelete: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5850&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a gray delete button component for triggering an action.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Poista',
    onClick: fn(),
    variant: 'gray-delete',
  },
};
export const RedDelete: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5850&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a red delete button component for triggering an action.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Poista',
    onClick: fn(),
    variant: 'red-delete',
  },
};

export const WhiteDelete: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5847&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a white delete button component for triggering an action.',
      },
    },
  },
  args: {
    label: 'Poista',
    onClick: fn(),
    variant: 'white-delete',
  },
};

export const Disabled: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5809&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a disabled button component.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Muokkaa',
    onClick: fn(),
    disabled: true,
  },
};

export const Small: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5831&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a small button component.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Muokkaa',
    onClick: fn(),
    size: 'sm',
  },
};

export const Large: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5857&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a large button component.',
      },
    },
  },
  args: {
    label: 'Muokkaa',
    onClick: fn(),
    size: 'lg',
    variant: 'white',
  },
};

export const MediumWithLeftIcon: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5837&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a medium button component with an left icon.',
      },
    },
  },
  args: {
    label: 'Takaisin',
    onClick: fn(),
    variant: 'white',
    icon: <MdArrowBack size={24} />,
  },
};

export const LargeWithRightIcon: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5853&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a large button component with an right icon.',
      },
    },
  },
  args: {
    label: 'Kokeile palvelua',
    onClick: fn(),
    size: 'lg',
    variant: 'white',
    icon: <MdArrowForward size={40} />,
    iconSide: 'right',
  },
};

export const AsLink: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-5869&t=6CqIVQA2gz0PPxiK-4',
    },
    docs: {
      description: {
        story: 'This is a link component that resembles the button component.',
      },
    },
    backgrounds,
  },
  args: {
    label: 'Kirjaudu sisään',
    size: 'lg',
    variant: 'gray',
    icon: <MdPerson size={40} />,
    iconSide: 'left',
    LinkComponent: ({ children }: { children: React.ReactNode }) => <a href="/#">{children}</a>,
  },
};
