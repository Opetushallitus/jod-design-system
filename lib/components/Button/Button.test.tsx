import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { Button } from './Button';

describe('Button', () => {
  it('renders the button with the correct label', () => {
    const label = 'Click me';
    const { container } = render(<Button label={label} onClick={vi.fn()} />);
    const button = screen.getByRole('button', { name: label });
    expect(button).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls the onClick function when clicked', async () => {
    const onClick = vi.fn();
    const { container } = render(<Button label="Click me" onClick={onClick} />);
    const user = userEvent.setup();
    const button = screen.getByRole('button', { name: 'Click me' });
    await user.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the button with the correct size', () => {
    const { container } = render(<Button label="Click me" onClick={vi.fn()} size="sm" />);
    const button = screen.getByRole('button', { name: 'Click me' });
    expect(button).toHaveClass('ds-text-button-sm');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the button with the correct variant', () => {
    const { container } = render(<Button label="Click me" onClick={vi.fn()} variant="gray" />);
    const button = screen.getByRole('button', { name: 'Click me' });
    expect(button).toHaveClass('ds-bg-bg-gray');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the button as disabled', () => {
    const { container } = render(<Button label="Click me" onClick={vi.fn()} disabled />);
    const button = screen.getByRole('button', { name: 'Click me' });
    expect(button).toBeDisabled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
