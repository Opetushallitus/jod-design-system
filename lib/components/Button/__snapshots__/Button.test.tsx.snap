// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > calls the onClick function when clicked 1`] = `
<button
  class="ds:flex ds:cursor-pointer ds:px-6 ds:rounded-[30px] ds:items-center ds:gap-4 ds:select-none ds:group ds:text-button-md ds:active:underline ds:focus-visible:underline ds:hover:not-disabled:underline ds:outline-offset-2 ds:disabled:cursor-not-allowed ds:bg-white ds:text-black ds:focus-visible:outline-secondary-1-dark ds:hover:text-secondary-1-dark ds:active:text-secondary-1-dark-2 ds:focus-visible:text-secondary-1-dark"
  type="button"
>
  <span
    class="ds:group-hover:underline ds:group-active:no-underline ds:group-focus-visible:no-underline ds:py-4"
  >
    Click me
  </span>
</button>
`;

exports[`Button > renders the button as disabled 1`] = `
<button
  class="ds:flex ds:cursor-pointer ds:px-6 ds:rounded-[30px] ds:items-center ds:gap-4 ds:select-none ds:group ds:text-button-md ds:active:underline ds:focus-visible:underline ds:hover:not-disabled:underline ds:outline-offset-2 ds:disabled:cursor-not-allowed ds:bg-white ds:text-inactive-gray"
  disabled=""
  type="button"
>
  <span
    class="ds:py-4"
  >
    Click me
  </span>
</button>
`;

exports[`Button > renders the button with the correct label 1`] = `
<button
  class="ds:flex ds:cursor-pointer ds:px-6 ds:rounded-[30px] ds:items-center ds:gap-4 ds:select-none ds:group ds:text-button-md ds:active:underline ds:focus-visible:underline ds:hover:not-disabled:underline ds:outline-offset-2 ds:disabled:cursor-not-allowed ds:bg-white ds:text-black ds:focus-visible:outline-secondary-1-dark ds:hover:text-secondary-1-dark ds:active:text-secondary-1-dark-2 ds:focus-visible:text-secondary-1-dark"
  type="button"
>
  <span
    class="ds:group-hover:underline ds:group-active:no-underline ds:group-focus-visible:no-underline ds:py-4"
  >
    Click me
  </span>
</button>
`;

exports[`Button > renders the button with the correct size 1`] = `
<button
  class="ds:flex ds:cursor-pointer ds:px-6 ds:rounded-[30px] ds:items-center ds:gap-4 ds:select-none ds:group ds:text-button-sm ds:active:underline ds:focus-visible:underline ds:hover:not-disabled:underline ds:outline-offset-2 ds:disabled:cursor-not-allowed ds:bg-white ds:text-black ds:focus-visible:outline-secondary-1-dark ds:hover:text-secondary-1-dark ds:active:text-secondary-1-dark-2 ds:focus-visible:text-secondary-1-dark"
  type="button"
>
  <span
    class="ds:group-hover:underline ds:group-active:no-underline ds:group-focus-visible:no-underline ds:py-2"
  >
    Click me
  </span>
</button>
`;

exports[`Button > renders the button with the correct variant 1`] = `
<button
  class="ds:flex ds:cursor-pointer ds:px-6 ds:rounded-[30px] ds:items-center ds:gap-4 ds:select-none ds:group ds:text-button-md ds:active:underline ds:focus-visible:underline ds:hover:not-disabled:underline ds:outline-offset-2 ds:disabled:cursor-not-allowed ds:bg-secondary-1-dark ds:active:bg-secondary-1-dark-2 ds:focus-visible:outline-secondary-1-dark ds:text-white"
  type="button"
>
  <span
    class="ds:group-hover:underline ds:group-active:no-underline ds:group-focus-visible:no-underline ds:py-4"
  >
    Click me
  </span>
</button>
`;
